tests/app/test_message.py

```diff
@@ -3,51 +3,127 @@
+        assert "No coil or register found to match 'bad_register'" in str(ex)
+        with pytest.raises(UnknownCommandError) as ex:
+            CommandMessage("", False, self.configuration)
+        assert "No coil or register found" in str(ex)
+
+    def test_bad_cmd_message_structure(self):
+        json_obj = {"action": "somecoil", "value": True}
+        for key in ("action", "value"):
+            json_obj["foo"] = json_obj[key]
+            del json_obj[key]
+            json_str = json.dumps(json_obj)
+            with pytest.raises(InvalidMessageError) as ex:
+                _ = CommandMessage.read(json_str)
+            assert "Message is missing required components" in str(ex.value)
+            assert ex.type == InvalidMessageError
+            json_obj[key] = json_obj["foo"]
+
+    def test_bad_cmd_message_syntax(self):
+        json_str = "not a real json string"
         with pytest.raises(InvalidMessageError) as ex:
             _ = CommandMessage.read(json_str)
-        assert "Message is missing required components" in str(ex.value)
+        assert "Message is invalid JSON syntax" in str(ex.value)
         assert ex.type == InvalidMessageError
-        json_obj[key] = json_obj["foo"]
+
+    def test_bad_cmd_message_datatype(self):
+        msg = CommandMessage("evgBatteryModeCoil", True, self.configuration)
+        try:
+            msg.validate()
+        except InvalidMessageError:
+            assert False
+        msg = CommandMessage("evgBatteryModeCoil", "foo", self.configuration)
+        with pytest.raises(InvalidMessageError) as ex:
+            msg.validate()
+        assert "The coil value 'foo' is invalid" in str(ex.value)
+        assert ex.type == InvalidMessageError
+
+
+class TestErrorMessage:
+    def test_good_err_message(self):
+        json_obj = {"category": "RuntimeError", "message": "oops"}
+        json_str = ErrorMessage.write(json_obj)
+        assert json_str == """{"category": "RuntimeError", "message": "oops"}"""
+
+    def test_bad_err_message(self):
+        json_obj = {
+            "category": "RuntimeError",
+            "message": "something went wrong",
+            "timestamp": datetime.now(),
+        }
+        with pytest.raises(InvalidMessageError) as ex:
+            _ = ErrorMessage.write(json_obj)
+        assert "serialised" in str(ex)
 
 
-def test_bad_cmd_message_syntax():
-    json_str = "not a real json string"
-    with pytest.raises(InvalidMessageError) as ex:
-        _ = CommandMessage.read(json_str)
-    assert "Message is invalid JSON syntax" in str(ex.value)
-    assert ex.type == InvalidMessageError
+class TestInputValidation:
+    def setup_method(self):
+        self.configuration = Configuration.from_file(
+            "tests/config/example_configuration.yaml"
+        )
 
+    def test_booleans(self):
+        # see https://docs.pydantic.dev/latest/usage/types/booleans/
+        pydantic_accepts = [
```

Comment:
I am cool with this. But if we decide to use all of these, we should also update the documentation of the config file, to explain that all of these are valid ways to modify coils.

@march83 I am keen to know your thoughts here.

Reply:
Yep, this works for me. I don't think there's anything obviously missing from this list

Reply:
Thanks @march83 . @serroba I've updated the example config file to include those values